---
title: "Methylome analysis"
subtitle: "ewas + neighb"
author: "Florent Chuffart, Ekatarina Flin, Fabien Jossaud"
date: "`r Sys.Date()`"
output: 
  rmarkdown::html_document:
    toc: true
    toc_float: true
    toc_depth: 3
    number_sections: true
---



```{r, echo=FALSE, eval=TRUE}
knitr::opts_chunk$set(collapse=TRUE, comment = "#>", fig.width=9, fig.height=6, eval=TRUE, echo=FALSE, results="hide", warning=FALSE)
source("../R/common.R")
source("config")
if (file.exists("00_local_config")) {source("00_local_config")}
# source("modelcall.R")
```


```{r inputs}
if (!exists("study_filename"))  study_filename = "study_GSE42861.rds"   ;
if (!exists("model_formula"))   model_formula = "meth~age"              ;
if (!exists("model_func_name")) model_func_name = "modelcalllm"         ;
if (!exists("newas"))           newas = 2000                            ;
if (!exists("neighb"))          neighb = 1000                           ;

if (file.exists(paste0(model_func_name, ".R"))) {
  source(paste0(model_func_name, ".R"))
}
```

```{r outputs}
options(scipen=999)
bed_ewas_filename = paste0("ewas4combp_", study_filename, "_", model_func_name, "_", model_formula, ".bed")
gse = substr(study_filename, 7, nchar(study_filename)-4)
prefix = paste0(gse, "_", model_func_name, "_", model_formula, "_ewas", newas, "_nn", neighb)
study_output_filename = paste0("study_", prefix, ".rds")
df_output_filename = paste0("df_", prefix, ".rds")
```

```{r loading data}
print(paste0("Loading ", study_filename, "..."))
s = mreadRDS(study_filename)
```

# Objective

The main objective of this vignette is to reduce the number of probes of the study `r study_filename` : 

  - we perform an ewas on the dataset using the model : `r model_formula` (computed by `r model_func_name`).
  - ewas results are exported as: [`r bed_ewas_filename`](./`r bed_ewas_filename`)
  - we select `r newas` probes with the lowest p.value (Fisher).
  - we iteratively add probes that own to the neighborhood of selected probes (+/- `neighb`b).
  - isolated probes are exported as : [`r study_output_filename`](./`r study_output_filename`)
  - new study is created as: [`r study_output_filename`](./`r study_output_filename`)
 
model is caracterised by the cofactor `r cofact = rownames(attr(stats::terms(as.formula(model_formula)), "factor"))[2]; print(cofact)` taking is values in `r levels(s$exp_grp[,cofact])`.

Final goal is to identify the Health Predictor Methylated Probes (HPMP) as well as Health Predictor Methylated Regions (HPMR). 


# Method 

The analysis of methylation is performed in two main steps. 

First, we realized the epigenome-wide association study (EWAS) to identify the differentially methylated individual sites. This method aims to calculate a linear regression model between the selected parameters `r model_formula`, and to obtain the corresponding p-values. The distribution of the p-values is then compared to the uniform distribution.

In the second step, ...

**TODO**

# Data

We use the methylation data serialized in the file `r study_filename`, containing `r nrow(s$exp_grp)` samples caracterized by their Illumina bead chip methylome (`r nrow(s$platform)` methylation sites with annotations).

```{r data, results="verbatim"}
if (FILTER_SEXUAL_CHR) {
  dim(s$data)
  s$platform = s$platform[!s$platform[,1]%in%c("chrX","chrY"),]
  s$data = s$data[rownames(s$platform),]  
  dim(s$data)
} 
d = s$data
e = s$exp_grp
pf = s$platform


pf_orig = pf
pf_chr_colname = colnames(pf_orig)[1]
pf_pos_colname = colnames(pf_orig)[2]

print("# dim meth matrix:")
dim(d)

print("# experimental grouping:")
head(e)
dim(e)

for (f in rownames(attr(stats::terms(as.formula(model_formula)), "factor"))[-1]) {
  if (is.character(e[[f]])) {
    print(paste0("Converting ", f, " as factor."))
    e[[f]] = as.factor(e[[f]])
  }
}


layout(1, respect=TRUE)
for (f in rownames(attr(stats::terms(as.formula(model_formula)), "factor"))[-1]) {
  if (is.factor(e[[f]])) {
    table(e[[f]], useNA="ifany")    
  } else if (is.numeric(e[[f]])) {
    plot(density(e[[f]], na.rm=TRUE), main=f)
  } else {
    stop(paste0("Covariate ", f, " is not factor nor numeric."))
  }
}
```


# Identification of differentially methylated sites using EWAS


`r model_formula`


```{r ewas}
USE_PARAPPLY = TRUE
if (length(rownames(attr(stats::terms(as.formula(model_formula)), "factor"))) < 3) {
  nb_fact_of_interest = 1
} else if (rownames(attr(stats::terms(as.formula(model_formula)), "factor"))[-1][2] == "hb") {
  nb_fact_of_interest = 2
} else {
  nb_fact_of_interest = 1  
}
ewas_orig = ewas = mewas_func2(d=d, e=e, USE_PARAPPLY=USE_PARAPPLY, model_formula=model_formula, model_func_name=model_func_name, nb_fact_of_interest=nb_fact_of_interest)
if (nrow(d)==nrow(ewas)) {
  rownames(ewas) = rownames(ewas_orig) = rownames(d)  
} else {
  stop(paste0("ewas do not have the same dimension as d."))
}

dim(ewas)
ewas = ewas[!is.na(ewas[,"lpv"]),]
dim(ewas)


if (FILTER_EWAS_FOR_DMR_CANDIDATE) {
  dmr_candidates = mbuild_dmr_candidates(pf=pf, pf_chr_colname="chr", pf_pos_colname="pos")
  dmr_candidates = dmr_candidates[sapply(dmr_candidates, length) > 1]
  idx_probes = unique(unlist(dmr_candidates))
  idx_probes = intersect(idx_probes, rownames(ewas))
} else {
  idx_probes = rownames(ewas)
}

ewas = ewas[idx_probes,]

dim(ewas)
```

```{r ewas_fabien, eval=FALSE}
y_key = "age"
confounders = NULL

if (nbewasprobes>0) {
  print("filtering EWAS top probes")
  cpg_matrix = d
  if (!is.null(confounders)) {
    design_matrix <- model.matrix(formula(paste0("~", y_key, "+", paste0(confounders, collapse="+"))), e)    
  } else {
    design_matrix <- model.matrix(formula(paste0("~", y_key)), e)    
  }
  dim(design_matrix)
  dim(cpg_matrix)
  fit = mlimma_lmFit(cpg_matrix, design_matrix)
  fit = limma::eBayes(fit) # compute moderated t-statistics, moderated F-statistic, and log-odds of differential expression by empirical Bayes moderation of the standard errors towards a global value
  # head(fit$p.value)
  idx_cpg_oi = rownames(fit$p.value)[rank(fit$p.value[,2]) <= nbewasprobes]
  layout(1, respect=TRUE)
  plot(fit$coefficients[,2], -log10(fit$p.value[,2]), col=(rownames(fit$p.value) %in% idx_cpg_oi)+1, pch=".", main=paste0(nbewasprobes, " probes"))

  litterature_models = readRDS("litterature_models.rds")  
  idx_cpg_litt = unique(unlist(lapply(lapply(litterature_models, "[[", "coeff"), "[[", "probes")))  

  idx_cpg = intersect(idx_cpg, c(idx_cpg_oi, idx_cpg_litt))

  dim(df)
  df_ewas = df[,c(idx_clinicals, idx_cpg)]
  dim(df_ewas)


  # df_dnamaging = df
  # colnames(df_dnamaging)[2] = "age"
  # head(df_dnamaging[,1:20])
  # save(df_dnamaging, file='./dnamaging/data/df_dnamaging.RData' , compress='xz')
} else {
  # Nothing to do for idx_cpg
}
```








```{r plot_ewas}
layout(matrix(1:4, 2, byrow=TRUE), respect=TRUE,)
plot(ewas[,"beta"], ewas[,"lpv"], main="Volcano plot", pch=".")
plot(density(ewas[,"lpv"]), main="lpv distribution", pch=".", xlim=c(0,100))
plot(density(10^-ewas[,"lpv"]), main="pv distribution", pch=".", xlim=c(0,1))
plot(sort(-log10(runif(nrow(ewas)))) , sort(ewas[,"lpv"]), main="million dollar plot", pch=".")
abline(a=0,b=1, lty=2, col="grey")
```

```{r build_bed}
head(ewas)
head(pf[,1:6])
lpval = ewas[,"lpv"]

pval = 10^-lpval

# build and write bed file
bed = pf[rownames(ewas),1:2] # Warning, here pf is global, it must be arg and indexed!!
head(bed)
bed[,3] = bed[,2]+1
bed$probes = names(pval)
bed$pval = pval
bed$strand = "+"
colnames(bed) = c("#chrom", "start", "end", "probes", "pval", "strand")
head(bed)

# write ewas res for combp
bed[,1] = as.character(bed[,1])
bed = bed[order(bed[,1], bed[,2]),]
write.table(bed,file=bed_ewas_filename , sep="\t", quote=FALSE, row.names=FALSE, col.names=TRUE)
# bed_on_disk = read.table(bed_ewas_filename, sep="\t", header=TRUE)
```

EWAS results is exporetd as: 
[`r bed_ewas_filename`](./`r bed_ewas_filename`)


# Selection of probes with the lowest Fisher p.value


```{r newas}
# bed_on_disk = read.table(bed_ewas_filename, sep="\t", header=TRUE)
# head(bed_on_disk)
dim(bed)
head(bed[order(bed$pval),])
tail(bed[order(bed$pval),])
idx_cpg_newas = rownames(bed[order(bed$pval),])[1:newas]


# HERE ALGO NN
if (FALSE) {
  platform = bed
  platform = platform[order(platform$chr,platform$pos),]
  platform$pos_prev_nn = c(NA,platform$start[-length(platform$start)])
  platform$pos_next_nn =  c(platform$start[-1],NA)
  platform$dist_prev_nn = platform$start - platform$pos_prev_nn
  platform$dist_next_nn = platform$pos_next_nn - platform$start
  platform[is.na(platform$dist_prev_nn) | platform$dist_prev_nn < 0,]$dist_prev_nn = 2*neighb
  platform[is.na(platform$dist_next_nn) | platform$dist_next_nn < 0,]$dist_next_nn = 2*neighb
  platform$is_kept = FALSE
  platform[idx_cpg_newas,]$is_kept = TRUE
  new_count = sum(platform$is_kept)
  old_count = new_count -1 
  while(old_count < new_count) {
    old_count = new_count
    platform$is_kept = platform$is_kept | (platform$dist_prev_nn <= neighb & c(FALSE,platform$is_kept[-length(platform$is_kept)] %in%TRUE))
    print(sum(platform$is_kept))
    platform$is_kept = platform$is_kept | (platform$dist_next_nn <= neighb & c(platform$is_kept[-1],FALSE) %in%TRUE)
    new_count = sum(platform$is_kept)
  }
  idx_cpg_newas_neighb_nn = rownames(platform)[platform$is_kept]  
  length(idx_cpg_newas_neighb_nn)
  length(idx_cpg_newas_neighb)
  sum(idx_cpg_newas_neighb_nn %in% idx_cpg_newas_neighb)
  sum(idx_cpg_newas_neighb %in% idx_cpg_newas_neighb_nn)
  head(bed[idx_cpg_newas_neighb[!idx_cpg_newas_neighb %in% idx_cpg_newas_neighb_nn],])

  bed$foo = NA
  bed$bar = NA
  bed[idx_cpg_newas_neighb, ]$foo =   idx_cpg_newas_neighb %in% idx_cpg_newas_neighb_nn
  bed[idx_cpg_newas_neighb, ]$bar =   idx_cpg_newas_neighb %in% idx_cpg_newas
  head(bed[idx_cpg_newas_neighb, ], 100)
  
  fat_feat[fat_feat$chr%in%"chr1",]
  fat_feat_indexed_probes[["chr1:1549615-1567821"]] %in% idx_cpg_newas
  fat_feat_indexed_probes[["chr1:1549615-1567821"]][fat_feat_indexed_probes[["chr1:1549615-1567821"]] %in% idx_cpg_newas]

  bed["cg08231710", ]
  bed[fat_feat_indexed_probes[["chr1:1549615-1567821"]],]
}

# HERE ALGO FAT_FEAT
pf = bed
pf_chr_colname = 1 
pf_pos_colname = 2
extend_region_dist = neighb/2 - 1

pf = pf[pf[,pf_pos_colname]>0,]
pf = pf[order(pf[[pf_chr_colname]],pf[[pf_pos_colname]]), ]
## index meth probes by chr
chrs = unique(pf[[pf_chr_colname]])
chrs_indexed_methpf = lapply(chrs, function(chr) {
    print(chr)
    idx = rownames(pf)[!is.na(pf[[pf_chr_colname]]) & pf[[pf_chr_colname]]==chr]
    ret = pf[idx,]
    return(ret)
})
names(chrs_indexed_methpf) = chrs

fat_feat = lapply(unique(pf[,pf_chr_colname]), function(chr) {
  d = pf[pf[,pf_chr_colname]==chr,c(pf_chr_colname, pf_pos_colname)]
  i = intervals::Intervals(c(d[,2], d[,2]+1), type="Z")
  # enlarge your fat feat
  l = extend_region_dist
  c = intervals::close_intervals( intervals::contract( intervals::reduce(intervals::expand(i, l)), l) )
  dim(c)
  df = data.frame(chr, c[,1], c[,2])
  return(df)
})
fat_feat = do.call(rbind, fat_feat)
dim(fat_feat)
fat_feat[,4] = paste0(fat_feat[,1], ":", fat_feat[,2], "-", fat_feat[,3])
fat_feat[,5] = fat_feat[,3] - fat_feat[,2]
fat_feat[,6] = "+"
# fat_feat = fat_feat[fat_feat[,5]>1,]
rownames(fat_feat) = fat_feat[,4]
colnames(fat_feat)  = c("chr", "start", "end", "id", "score", "strand")
dim(fat_feat)
head(fat_feat)

## index probes by feat name
print("# indexing probes by feat name")
fat_feat_indexed_probes = epimedtools::monitored_apply(fat_feat, 1, function(feat) {
  # feat = fat_feat[3,]
  # print(feat)
  chr = feat[[1]]
  len = as.numeric(feat[[5]])
  meth_platform = chrs_indexed_methpf[[chr]]
  ret = dmprocr_get_probe_names(feat, meth_platform, pf_chr_colname, pf_pos_colname, 0, len)
  # meth_platform[ret,1:3]
  # feat
  return(ret)
})

fat_feat$n_probes = sapply(fat_feat_indexed_probes, length) 
fat_feat$z_sidak_p = 1


layout(matrix(1:2,1), respect=TRUE)
barplot(table(cut(fat_feat$score, breaks=c(0, 100, 200, 300, 500, 1000, 2000, 3000, 5000, 10000, 20000, 30000, 50000, 100000, 200000, 50000000), include.lowest=TRUE)), las=2,main="locus length distribution")
barplot(table(cut(fat_feat$score[fat_feat$score>1], breaks=c(0, 100, 200, 300, 500, 1000, 2000, 3000, 5000, 10000, 20000, 30000, 50000, 100000, 200000, 50000000), include.lowest=TRUE)), las=2,main="locus length distribution")


idx_fat_feat_oi = sapply(fat_feat_indexed_probes, function(probes) {
  any(probes %in% idx_cpg_newas)
})
fat_feat = fat_feat[names(idx_fat_feat_oi)[idx_fat_feat_oi],]
fat_feat_indexed_probes = fat_feat_indexed_probes[rownames(fat_feat)]
idx_cpg_newas_neighb = unlist(unique(fat_feat_indexed_probes))
length(idx_cpg_newas_neighb)

barplot(table(cut(fat_feat$score, breaks=c(0, 100, 200, 300, 500, 1000, 2000, 3000, 5000, 10000, 20000, 30000, 50000, 100000, 200000, 50000000), include.lowest=TRUE)), las=2,main="locus length distribution")
barplot(table(cut(fat_feat$score[fat_feat$score>1], breaks=c(0, 100, 200, 300, 500, 1000, 2000, 3000, 5000, 10000, 20000, 30000, 50000, 100000, 200000, 50000000), include.lowest=TRUE)), las=2,main="locus length distribution")

plot(sort(-log10(runif(nrow(ewas)))) , sort(ewas[,"lpv"]),                              ylim=c(0, max(ewas[,"lpv"])), main="million dollar plot", pch=".", col=(names(sort(ewas[,"lpv"]))%in%idx_cpg_newas)+1)
abline(a=0,b=1, lty=2, col="grey")
legend("topleft", pch=1, col=2, paste0(length(idx_cpg_newas), " lowest pvals"))
plot(sort(-log10(runif(nrow(ewas)))) , sort(ewas[,"lpv"]),                              ylim=c(0, max(ewas[,"lpv"])), main="million dollar plot", pch=".", col=(names(sort(ewas[,"lpv"]))%in%idx_cpg_newas_neighb)+1)
abline(a=0,b=1, lty=2, col="grey")
legend("topleft", pch=1, col=2, paste0(length(idx_cpg_newas_neighb), " probes in ", newas, " lowest\n ewas probes +/-", neighb, " bp"))

# stop("EFN")

s$data = s$data[rownames(s$data)%in%idx_cpg_newas_neighb,]  
s$platform = s$platform[rownames(s$data),]
saveRDS(s, study_output_filename)


df = cbind(s$exp_grp, t(s$data[,rownames(s$exp_grp)]))
print(paste0("Writing ", df_output_filename, "..."))
saveRDS(df, df_output_filename)
dim(df)



```



```{r combp_outputs}
# # recuparation du fichier dmrbycombp1000_SGCE.fdr.bed après comb-p
# combp_res_region_filename = paste0("dmr_", prefix, ".regions-t.bed")
# combp_res_region_filename = "dmr_study_r0_ewas2000_nn1000_GSE42861.rds_modelcalllm_meth~age+gender_1e-30.regions-t.bed"
# combp_res_region = read.table(combp_res_region_filename, comment="@", header=TRUE)
#   X.chrom    start      end     min_p n_probes       z_p z_sidak_p
# 1    chr1  2344979  2345476 0.000e+00       11 4.676e-47 2.025e-45
# 2    chr1  3567870  3567931 1.230e-31        2 6.901e-18 2.435e-15
# 3    chr1 17766917 17766996 3.845e-31        2 1.070e-32 2.915e-30
# 4    chr1 19992504 19992565 0.000e+00        2 3.848e-44 1.358e-41
# 5    chr1 39491318 39491460 6.332e-31        3 1.852e-32 2.808e-30
# 6    chr1 47882314 47882740 0.000e+00        4 7.368e-48 3.722e-46
# combp_res_region = data.frame(n_probes=NA, z_sidak_p=NA)[-1,]
# results = combp_res_region
results = fat_feat
results[,5] = NA
results[,6] = NA
names(results)[5:6] = c("n_probes", "z_sidak_p")
tmp_results_filename = paste0("tmp_", prefix, ".bed")
results_filename = paste0("global_results_", prefix, ".xlsx")
```







# Annotation




```{r, eval=nrow(results)>0}
print("# # annotate dmr")
rownames(results) = paste0(results[,1], ":", results[,2], "-", results[,3])
results = cbind(results[,1:3], id=rownames(results), len=results[,3] - results[,2], strand="+", results[,4:ncol(results)])
head(results)
```


```{r dmr_indexed_probes, eval=nrow(results)>0}
print("# # dmr_indexed_probes")
# meth pf
# pf_chr_colname = "seqnames"
# pf_pos_colname = "start"
# pf_orig = mreadRDS("~/projects/datashare/platforms/EPIC.hg38.manifest.full.fch.rds")
pf_orig = pf_orig[pf_orig[,pf_pos_colname]>0,]    
pf_orig = pf_orig[order(pf_orig[[pf_chr_colname]],pf_orig[[pf_pos_colname]]), ]
## index meth probes by chr
chrs = unique(pf_orig[[pf_chr_colname]])
chrs_indexed_methpf = lapply(chrs, function(chr) {
  # print(chr)
  idx = rownames(pf_orig)[!is.na(pf_orig[[pf_chr_colname]]) & pf_orig[[pf_chr_colname]]==chr]  
  ret = pf_orig[idx,]
  return(ret)
})
names(chrs_indexed_methpf) = chrs

dmr_indexed_probes = epimedtools::monitored_apply(results, 1, function(feat) {
  # feat = fat_feat[3,]
  # print(feat)
  chr = feat[[1]]
  len = as.numeric(feat[[5]])
  meth_platform = chrs_indexed_methpf[[chr]]
  ret = dmprocr_get_probe_names(feat, meth_platform, pf_chr_colname, pf_pos_colname, 0, len) 
  # meth_platform[ret,1:3]
  # feat
  return(ret)
})
```


```{r effect_by_dmr_from_ewas, eval=nrow(results)>0}
results$ewas_pv     = NA
results$ewas_effect = NA
for (ff_id in rownames(results)) {
  if (nrow(results)>1 & !is.matrix(dmr_indexed_probes)) {
    p = intersect(dmr_indexed_probes[[ff_id]], rownames(ewas_orig))
  } else {
    p = intersect(dmr_indexed_probes[,ff_id], rownames(ewas_orig))    
  }
  results[ff_id, ]$ewas_pv     = signif(10^-mean(ewas_orig[p,2], na.rm=TRUE), 3)
  results[ff_id, ]$ewas_effect = signif(mean(ewas_orig[p,1], na.rm=TRUE), 3)    
}
head(results)
```



> as.vector(unlist(fat_feat_indexed_probes[rownames(fat_feat[fat_feat$score==1,])]))
 [1] "cg04080625" "cg17168836" "cg08893087" "cg00653615" "cg13804478"
 [6] "cg26894354" "cg08128734" "cg26575914" "cg06370069" "cg00533891"
[11] "cg19873536" "cg12776156" "cg15948836" "cg05655457" "cg12686055"
[16] "cg21656937" "cg11649376" "cg17183905" "cg23395688" "cg24674215"
[21] "cg06914505" "cg10577534" "cg06601581" "cg11024682" "cg21692159"
[26] "cg12533658" "cg26748477" "cg06874016" "cg12350474" "cg13683374"
[31] "cg21469505" "cg24848615" "cg14950321" "cg25782440" "cg01542019"
[36] "cg26856607" "cg07197831" "cg10107473" "cg10835286" "cg06682024"
[41] "cg23482132" "cg20456243" "cg16640358" "cg19753867" "cg03758021"
[46] "cg17526301" "cg10803722" "cg06171420" "cg07127410" "cg18708252"
[51] "cg03530962" "cg22354646" "cg04357717" "cg00329615" "cg25425705"
 [56] "cg16782084" "cg06388350" "cg14069287" "cg13039251" "cg22788465"
[61] "cg21874213" "cg26403171" "cg09646173" "cg02114954" "cg16431787"
[66] "cg01882991" "cg17593342" "cg01205011" "cg13001142" "cg22133973"
[71] "cg17136255" "cg07568841" "cg09124496" "cg25311470" "cg01493685"
[76] "cg00582628" "cg25998745" "cg04436528" "cg00755588"

```{r features, eval=TRUE}
# https://zenbu-wiki.gsc.riken.jp/zenbu/wiki/index.php/Uploading_UCSC_repetitive_elements_track
# downloading UCSC rmsk data as BED
# BED formatted UCSC track content can be obtained from UCSC table broswer.
# The rmsk RepeatMasker (rmsk) track can be exported as BED file by selecting
#
# the assembly "Feb.2009 GRCh37/hg19"
# the group "repeats and variations"
# the track "RepeatMasker"
# and finally the table "rmsk"
# As we desire the complete repetitive elements genome-wide to be loaded into ZENBU, therefore we select
#
# region: "genome"
# ZENBU enable gzip compressed bed files to be loaded directly, so we will further select :
#
# output format: "BED - browser extensible data"
# output file: we will name the file "UCSC_rmsk.hg19.bed.gz"
# file type returned: "gzip compressed"

if (!exists("rmsk_raw")) {
  rmsk_raw = read.table(gzfile("UCSC_rmsk.hg19.bed.gz"), skip=1, stringsAsFactors=FALSE)
}
rmsk = rmsk_raw
# head(rmsk)
# dim(rmsk)
# head(rmsk[,4])
#
# foo = rmsk[grep("AATGG|ATGGA|ATTCC|CATTC|GAATG|TCCAT|TGGAA", rmsk[,4]),]
# bar = rmsk[grep("ALR/Alpha", rmsk[,4]),]
#
# foo = rmsk[grep("line", rmsk[,4], ignore.case=TRUE),]
# table(foo[,4])
# foo = rmsk[grep("sine", rmsk[,4], ignore.case=TRUE),]
# table(foo[,4])
# foo = rmsk[grep("alu", rmsk[,4], ignore.case=TRUE),]
# table(foo[,4])
if (!exists("annotations_rmsk")) {
  # annots = c("mm10_lncrna_gencode")
  # annots = c('mm10_cpgs', 'mm10_basicgenes', "mm10_lncrna_gencode", "mm10_enhancers_fantom")
  # annots = c('mm10_cpgs', 'mm10_basicgenes', "mm10_enhancers_fantom")
  # annots = c( 'mm10_basicgenes', 'mm10_cpgs')
  # write.table(rmsk[rmsk[,4]=="GSAT_MM",]  , file="gsat.mm10.bed"  , sep="\t", quote=FALSE,row.names=FALSE, col.names=FALSE)
  # write.table(rmsk[rmsk[,4]=="SYNREP_MM",], file="synrep.mm10.bed", sep="\t", quote=FALSE,row.names=FALSE, col.names=FALSE)
  write.table(rmsk[grep("AATGG|ATGGA|ATTCC|CATTC|GAATG|TCCAT|TGGAA", rmsk[,4]),], file="ggaat.hg19.bed"  , sep="\t", quote=FALSE,row.names=FALSE, col.names=FALSE)
  write.table(rmsk[grep("ALR/Alpha", rmsk[,4]),]                                , file="alralpha.hg19.bed", sep="\t", quote=FALSE,row.names=FALSE, col.names=FALSE)
  write.table(rmsk[grep("line", rmsk[,4], ignore.case=TRUE),]                   , file="line.hg19.bed", sep="\t", quote=FALSE,row.names=FALSE, col.names=FALSE)
  write.table(rmsk[grep("sine", rmsk[,4], ignore.case=TRUE),]                   , file="sine.hg19.bed", sep="\t", quote=FALSE,row.names=FALSE, col.names=FALSE)
  write.table(rmsk[grep("alu", rmsk[,4], ignore.case=TRUE),]                    , file="alu.hg19.bed" , sep="\t", quote=FALSE,row.names=FALSE, col.names=FALSE)
  annotatr::read_annotations(con="ggaat.hg19.bed"  , genome = "hg19", name = "ggaat"  , format = "bed")
  annotatr::read_annotations(con="alralpha.hg19.bed", genome = "hg19", name = "alralpha", format = "bed")
  annotatr::read_annotations(con="line.hg19.bed", genome = "hg19", name = "line", format = "bed")
  annotatr::read_annotations(con="sine.hg19.bed", genome = "hg19", name = "sine", format = "bed")
  annotatr::read_annotations(con="alu.hg19.bed" , genome = "hg19", name = "alu", format = "bed")
  annots = c( "hg19_custom_ggaat", "hg19_custom_alralpha", "hg19_custom_line", "hg19_custom_sine", "hg19_custom_alu")
  # annots = c( "hg19_basicgenes", "hg19_cpgs", "hg19_custom_gsat", "hg19_custom_synrep")
  # annots = "hg19_basicgenes"
  # annots = "hg19_cpgs"
  annotations_rmsk = annotatr::build_annotations(genome = 'hg19', annotations = annots)  
}



head(s$platform[,1:6])
dim(s$platform)
foo = s$platform[,1:6]
foo[,6] = foo[,3]
foo[,3] = foo[,2]+1
head(foo)
tmp_results_filename = "tmp_pf.bed"
print(tmp_results_filename)
write.table(foo, file=tmp_results_filename, sep="\t", quote=FALSE,row.names=FALSE, col.names=FALSE)      
probes_regions = annotatr::read_regions(con=tmp_results_filename, genome=genome, format="bed")

probes_annotated = annotatr::annotate_regions(
    regions = probes_regions,
    annotations = annotations_rmsk,
    ignore.strand = TRUE,
    quiet = FALSE)
# A GRanges object is returned
df_probes_annotated = data.frame(probes_annotated)
# dedup
# df_probes_annotated$annot.type = factor(df_probes_annotated$annot.type, levels=paste0(genome, c("_genes_promoters", "_genes_1to5kb", "_genes_5UTRs", "_genes_exons", "_genes_introns", "_genes_3UTRs")))
table(df_probes_annotated$annot.type)
df_probes_annotated$name
```











```{r annotate dmr, eval=nrow(results)>0}
# annotatr
print(tmp_results_filename)
write.table(results[,1:6], file=tmp_results_filename, sep="\t", quote=FALSE,row.names=FALSE, col.names=FALSE)      
dm_regions = annotatr::read_regions(con=tmp_results_filename, genome=genome, format="bed")
# Build the annotations (a single GRanges object)
if (!exists("annotations_list")) {
  annotations_list = list()
}
if (! genome %in% names(annotations_list)) {
  # annots = paste0(genome, c("_lncrna_gencode"))
  # annots = paste0(genome, c('_cpgs', '_basicgenes', "_lncrna_gencode", "_enhancers_fantom"))
  # annots = paste0(genome, c('_cpgs', '_basicgenes', "_enhancers_fantom"))
  # annots = paste0(genome, c('_cpgs', '_basicgenes'))
  annots = paste0(genome, c("_basicgenes"))
  annotations_list[[genome]] = annotatr::build_annotations(genome=genome, annotations=annots)  
}
annotations = annotations_list[[genome]]
# Intersect the regions we read in with the annotations
dm_annotated = annotatr::annotate_regions(
    regions = dm_regions,
    annotations = annotations,
    ignore.strand = TRUE,
    quiet = FALSE)
# A GRanges object is returned
df_dm_annotated = data.frame(dm_annotated)
# dedup
df_dm_annotated$annot.type = factor(df_dm_annotated$annot.type, levels=paste0(genome, c("_genes_promoters", "_genes_1to5kb", "_genes_5UTRs", "_genes_exons", "_genes_introns", "_genes_3UTRs")))
table(df_dm_annotated$annot.type)
table(df_dm_annotated$name)
df_dm_annotated = df_dm_annotated[order(df_dm_annotated$name, df_dm_annotated$annot.type),]
df_dm_annotated = df_dm_annotated[!duplicated(paste(df_dm_annotated$name, df_dm_annotated$annot.symbol)),]
dim(df_dm_annotated)
table(df_dm_annotated$annot.type)
# aggregat by fat_feat
fat_feat_annotations = lapply(unique(df_dm_annotated$name), function(n) {
  # n = "chr7:2661785-2662169"
  foo = df_dm_annotated[df_dm_annotated$name==n,]
  if (sum(df_dm_annotated$name==n) == 1) {
    return(foo)
  } else {
    foo[["annot.symbol"]] = paste0(foo[["annot.symbol"]],collapse=";")
    foo[["annot.type"]] = paste0(foo[["annot.type"]],collapse=";")
    foo[["annot.tx_id"]] = paste0(foo[["annot.tx_id"]],collapse=";")
    return(foo[1,])
  }
})
fat_feat_annotations = do.call(rbind, fat_feat_annotations)
rownames(fat_feat_annotations) = fat_feat_annotations$name
head(fat_feat_annotations)
rownames(fat_feat_annotations) %in% rownames(results)
rownames(results) %in% rownames(fat_feat_annotations)
# inject into results
results$annot.type   = NA
results$annot.symbol = NA
results$annot.tx_id  = NA
for (idx in rownames(fat_feat_annotations)) {
  results[idx,"annot.type"  ] = as.character(fat_feat_annotations[idx,"annot.type"  ])
  results[idx,"annot.symbol"] = as.character(fat_feat_annotations[idx,"annot.symbol"])
  results[idx,"annot.tx_id" ] = as.character(fat_feat_annotations[idx,"annot.tx_id" ])  
}

results$nb_probes = NA
results$probes = NA


for (idx in rownames(results)) {
  if (nrow(results)>1 & !is.matrix(dmr_indexed_probes)) {
    idx_probes = dmr_indexed_probes[[idx]]
  } else {
    idx_probes = dmr_indexed_probes[,idx]
  }
  results[idx, "nb_probes"] = length(idx_probes)
  results[idx, "probes"] = paste(idx_probes, collapse=";")
}

tmpidx = order(results[,"z_sidak_p"])
results = results[tmpidx,]

# results$url = sapply(results[,4], function(i){
#   output_pdf_filename = paste0("global_results/", prefix, "_", i, ".pdf")
#   paste0("http://epimed.univ-grenoble-alpes.fr/downloads/florent/expedition_5300/results/meth/", output_pdf_filename)
# })

print(results_filename)
WriteXLS::WriteXLS(results, results_filename, FreezeCol=6, FreezeRow=1, BoldHeaderRow=TRUE, AdjWidth=TRUE)
```


























# Results

`r study_filename`

$$\text{`r model_formula`}$$

There are `r nrow(results)` regions of interest with `newas`=$`r newas`$ and `neighb`=$`r neighb`$.

[`r results_filename`](./`r results_filename`)

```{r results="verbatim", eval=nrow(results)>0}
tmp_results = results
rownames(tmp_results) = NULL
nb_max_res = 12
print(tmp_results[1:min(nb_max_res, nrow(tmp_results)),])
```


```{r, eval=nrow(results)>0, eval=FALSE}
# combp_res_probes = read.table(gzfile(paste0("dmr_study_r0_ewas2000_nn1000_GSE42861.rds_modelcalllm_meth~age+gender_1e-30.fdr.bed.gz")), comment="@", header=TRUE)
# head(combp_res_probes)
combp_res_probes = bed
combp_res_probes[,4]= combp_res_probes[,5]
combp_res_probes[,6]= combp_res_probes[,5]
head(combp_res_probes)

df_ewas = data.frame(ewas)
head(df_ewas)
df_ewas[,4] = rownames(df_ewas)
head(df_ewas)

if (nrow(results) > 0) {
  for (i in rownames(results)[1:min(nb_max_res, nrow(results))]) {
    print(i)
    roi = results[i,]
    if (nrow(results)>1 & !is.matrix(dmr_indexed_probes)) {
      idx_probes = dmr_indexed_probes[[i]]
    } else {
      idx_probes = dmr_indexed_probes[,i]
    }
    plot_res(
      roi, 
      combp_res_probes=combp_res_probes, 
      ewas=df_ewas, 
      pheno_key=rownames(attr(stats::terms(as.formula(model_formula)), "factor"))[2],
      idx_probes=idx_probes,
      study_filename=study_filename
    )
  } 
}
```



# Session Information

```{r, results="verbatim"}
sessionInfo()
```




